src/tls/cockpit-certificate-helper.in

```diff
@@ -2,51 +2,88 @@
+        chown root:"${COCKPIT_GROUP}" "${DESTINATION}"
+        chmod g+r "${DESTINATION}"
+    fi
+}
 
-# ensure this gets run with a non-C locale; ipa fails otherwise
-if [ $(sh -c 'eval `locale`; echo $LC_CTYPE') = 'C' ]; then
-    export LC_CTYPE=C.UTF-8
-fi
+ipa() {
+    local REALM="$1"
+    local USER="$2"
 
-# create a kerberos Service Principal Name for cockpit-ws, unless already present
-ipa service-show "${SERVICE}" || ipa service-add --ok-as-delegate=true --force "${SERVICE}"
+    local HOST="$(hostname -f)"
+    local SERVICE="HTTP/${HOST}@${REALM}"
 
-# add cockpit-ws key, unless already present
-mkdir -p /etc/cockpit
-klist -k /etc/cockpit/krb5.keytab | grep -qF "${SERVICE}" || ipa-getkeytab -p "HTTP/${HOST}" -k /etc/cockpit/krb5.keytab
+    # use a temporary keytab to avoid interfering with the system one
+    export KRB5CCNAME=/run/cockpit/keytab-setup
 
-# ipa-getcert cannot set permisisons nor write into /etc/cockpit due to SELinux
-if ipa-getcert request -f /run/cockpit/ipa.crt -k /run/cockpit/ipa.key -K "HTTP/${HOST}" -w -v; then
-    mv -Z /run/cockpit/ipa.crt /etc/cockpit/ws-certs.d/10-ipa.cert
-    mv -Z /run/cockpit/ipa.key /etc/cockpit/ws-certs.d/10-ipa.key
+    # ipa-getkeytab needs root to create the file, same for cert installation
+    [ $(id -u) = 0 ] || exit 0
 
-    # The certificate should be world-readable
-    chmod a+r /etc/cockpit/ws-certs.d/10-ipa.cert
+    # not an IPA setup? cannot handle this
+    type ipa >/dev/null 2>&1 || exit 0
 
-    # If COCKPIT_GROUP is set, then make sure the key is readable by that group too
-    if [ -n "${COCKPIT_GROUP}" ]; then
-        chown root:"${COCKPIT_GROUP}" /etc/cockpit/ws-certs.d/10-ipa.key
-        chmod g+r /etc/cockpit/ws-certs.d/10-ipa.key
+    # IPA operations require auth; read password from stdin to avoid quoting issues
+    # if kinit fails, we can't handle this setup, exit cleanly
+    kinit "${USER}@${REALM}" || exit 0
+
+    # ensure this gets run with a non-C locale; ipa fails otherwise
+    if [ $(sh -c 'eval `locale`; echo $LC_CTYPE') = 'C' ]; then
+        export LC_CTYPE=C.UTF-8
     fi
-fi
+
+    # create a kerberos Service Principal Name for cockpit-ws, unless already present
+    ipa service-show "${SERVICE}" || \
+        ipa service-add --ok-as-delegate=true --force "${SERVICE}"
+
+    # add cockpit-ws key, unless already present
+    klist -k /etc/cockpit/krb5.keytab | grep -qF "${SERVICE}" || \
+        ipa-getkeytab -p "HTTP/${HOST}" -k /etc/cockpit/krb5.keytab
+
+    # request the certificate (into the working directory)
+    ipa-getcert request -f 10-ipa.cert -k 10-ipa.key -K "HTTP/${HOST}" -w -v
+
+    # install it into /etc
+    install_cert 10-ipa.cert
+    install_key 10-ipa.key
+}
+
+main() {
+    # Create a private working directory
+    mkdir -p "${COCKPIT_RUNTIME_DIR}"
+    WORKDIR="${COCKPIT_RUNTIME_DIR}/certificate-helper"
+    mkdir -m 700 "${WORKDIR}" # we expect that not to have existed
+    trap 'rm -rf "${WORKDIR}"' EXIT
```

Comment:
I wonder how that works, given that WORKDIR is the pwd then? doesn't that need to `cd ..` (or someplace else) first?

Please also add `INT QUIT PIPE`, to also clean up on some common errors.

Reply:
I think there's no trouble with removing the current directory: particularly when you're right about to exit.

Reply:
Also: adding `INT QUIT PIPE` here doesn't really do the right thing: it seems like it causes those signals to be responded to by removing the working directory (instead of quitting the shell).

`set -e`, however, seems to cause the shell to quit immediately anyway (including with the `EXIT` trap).  *not* trapping `INT` causes the `EXIT` trap *not* to run, though.

So it seems like we probably need to set INT QUIT PIPE, etc. to `''` (no action) and keep the `EXIT` trap?  This is confusing and awful :rofl:

Reply:
I think the best way is to do something like this:

```sh
    trap 'exit' INT QUIT PIPE TERM
    trap 'rm -rf "${WORKDIR}"' EXIT
```

Reply:
Heh, that's not actually too bad then :)